spring-xd-messagebus-spi/src/main/java/org/springframework/xd/dirt/integration/bus/MessageBusSupport.java

```diff
@@ -898,7 +898,11 @@ private void revertDirectBindingIfNecessary(Binding binding) {
 
 		@Override
 		public int selectPartition(Object key, int partitionCount) {
-			return Math.abs(key.hashCode()) % partitionCount;
+			int hashCode = key.hashCode();
+			if (hashCode == Integer.MIN_VALUE) {
+				hashCode = 0;
+			}
+			return Math.abs(hashCode) % partitionCount;
```

Comment:
This fix was simply because `Math.abs(Integer.MIN_VALUE) == Integer.MIN_VALUE`.

I see your point but we can't change a strategy interface after a release candidate.

> What if partition is less than -meta.partitionCount ? The modulo should be always applied, then abs() IMHO

If it's already `0 >= partition < count` then it doesn't need to be modulo'd.

However, the module __should__ be applied if it's < 0 because the absolute value might be `> count`.

So I agree, but I think we're stuck with replacing your snippet with an unconditional modulo here while still passing it into the selector.